source/MaterialXGenShader/Nodes/SourceCodeNode.cpp

```diff
@@ -81,13 +81,44 @@ void SourceCodeNode::emitFunctionCall(const ShaderNode& node, GenContext& contex
         static const string prefix("{{");
         static const string postfix("}}");
 
-        // Inline expressions can only have a single output
+        size_t pos = 0;
+        size_t i = _functionSource.find_first_of(prefix);
+        int variableIndex = 1;
+        while (i != string::npos)
+        {
+            size_t j = _functionSource.find_first_of(postfix, i + 2);
+            if (j == string::npos)
+            {
+                throw ExceptionShaderGenError("Malformed inline expression in implementation for node " + node.getName());
+            }
+
+            const string variable = _functionSource.substr(i + 2, j - i - 2);
+            const ShaderInput* input = node.getInput(variable);
+            if (!input)
+            {
+                throw ExceptionShaderGenError("Could not find an input named '" + variable +
+                    "' on node '" + node.getName() + "'");
+            }
+
+            if (!input->connection)
+            {
+                string variableName = node.getName() + "_" + input->name + std::to_string(variableIndex++);
```

Comment:
Is the variableIndex counter needed to construct the variable name? I think node name and input name would be enough to make it locally unique.
To minimize the risk of having it collide with some other name globally we could add a constant postfix: "_tmp" or "_val" or something like that. For longer term we should probably consider having a variable naming system that guarantees the names are globally unique.

Reply:
I just realized, the same input {{foo}} can be given multiple times in the expression, so you also need to keep a std::set of which ones you have already defined.